app/domain/authentication/authn_k8s/web_socket_client.rb

```diff
@@ -26,16 +26,22 @@ def connect(url, options = {})
                                 uri.port || (uri.scheme == 'wss' ? 443 : 80))
         if %w[https wss].include?(uri.scheme)
           ctx = OpenSSL::SSL::SSLContext.new
-          ctx.ssl_version = options[:ssl_version] || 'SSLv23'
+          ssl_version = options[:ssl_version] || 'SSLv23'
+          ctx.ssl_version = ssl_version
           ctx.verify_mode = options[:verify_mode] || OpenSSL::SSL::VERIFY_NONE # use VERIFY_PEER for verification
           cert_store = options[:cert_store]
           unless cert_store
             cert_store = OpenSSL::X509::Store.new
             cert_store.set_default_paths
           end
           ctx.cert_store = cert_store
+
           @socket = ::OpenSSL::SSL::SSLSocket.new(@socket, ctx)
+          if ssl_version != 'SSLv23'
+            @socket.hostname = options[:hostname] || uri.host
+          end
           @socket.connect
+          @socket.post_connection_check(@socket.hostname) if @socket.hostname
```

Comment:
I would like to understand this better, since it's a bit more involved than my simple implementation of always setting the hostname.
1. Why are we only excluding this version ? Is it the case that SNI is only expected to be relevant on `ssl_version > SSLv23` ?
2. Can you explain the logic around the `post_connection_check`

Reply:
SNI is an extension in TLS only so we don't want to bother adding the hostname if they are using SSLv23 for some reason.

the post_connection_check will throw an error if we expected a certain SNI certificate but we didn't get it - this makes sure that if we are expecting SNI we get the proper certificate

Reply:
On board with (1). For (2) I was sure I was able to get errors thrown with just calls to `#connect`, I'll reconfirm.